[pythonscripting] Initial contribution (#18499)
openhab/openhab-addons: bundles/org.openhab.automation.pythonscripting/README.md

> # Python 3 Binding

Author: is fixed

This add-on provides support for Python 3 that can be used as a scripting language within automation rules.
It is based on [GraalPy](https://www.graalvm.org/python/) from the [GraalVM project](https://www.graalvm.org/).

> Also included is [openhab-python](https://github.com/HolgerHees/openhab-python), a python library to support automation in openHAB. It provides convenient access to common core openHAB functions that make the full range of Java APIs easily accessible and usable

Reviewer: ```suggestion
Also included is [openhab-python](https://github.com/HolgerHees/openhab-python), a python library to support automation in openHAB. It provides convenient access to common core openHAB functions that make the full range of Java APIs easily accessible and usable.
```

Author: fixed

[[toc]]

## Creating Python Scripts

When this add-on is installed, you can select Python3 as a scripting language when creating a script action within the rule editor of the UI.

Alternatively, you can create scripts in the `automation/python` configuration directory. If you create an empty file called `test.py`, you will see a log line with information similar to:

```text

> ... [INFO ] [ort.loader.AbstractScriptFileWatcher]] - (Re-)Loading script '/openhab/conf/automation/python/test.py'

Reviewer: ```suggestion
... [INFO ] [ort.loader.AbstractScriptFileWatcher] - (Re-)Loading script '/openhab/conf/automation/python/test.py'
```

Author: fixed

```

To enable debug logging, use the [console logging]({{base}}/administration/logging.html) commands to
enable debug logging for the automation functionality:

```text
log:set DEBUG org.openhab.automation.pythonscripting
```

## Scripting Basics

Lets start with a simple script

```python
from openhab import rule
from openhab.triggers import GenericCronTrigger

@rule( triggers = [ GenericCronTrigger("*/5 * * * * ?") ] )
class Test:
    def execute(self, module, input):
        self.logger.info("Rule was triggered")
```

or another one, using the [scope module](#module-scope)

```python
from openhab import rule
from openhab.triggers import ItemCommandTrigger

import scope

@rule( triggers = [ ItemCommandTrigger("Item1", scope.ON) ] )
class Test:
    def execute(self, module, input):
        self.logger.info("Rule was triggered")
```

::: tip Note
By default, the scope, Registry and logger is automatically imported for UI based rules
:::
 

> ## `PY3` Transformation

Reviewer: Once this is merged, we can update the UI provide doc links, code snippets etc, see https://github.com/openhab/openhab-webui/blob/main/bundles/org.openhab.ui/web/src/assets/transformations.js.

Reviewer: @HolgerHees Do you want to do this yourself?

openHAB provides several [data transformation services](https://www.openhab.org/addons/#transform) as well as the script transformations, that are available from the framework and need no additional installation.
It allows transforming values using any of the available scripting languages, which means Python Scripting is supported as well.

> See the [transformation docs](https://openhab.org/docs/configuration/transformations.html#script-transformation) for more general information on the usage of script transformations.

Reviewer 2: If I was to maintain it I would probably automate generating this README using some script (or Maven) so it will be easier to update the header and replace the URLs.

Note that line 20 is already using `{{base}}` which triggered me to comment on this.

> See the [transformation docs](https://openhab.org/docs/configuration/transformations.html#script-transformation) for more general information on the usage of script transformations.

Author: I remove the usage of {{base}} for now. But I understand you objections. I'll definitely work on a solution after the 5.0 release. For now, I'd like to keep things simple.

> See the [transformation docs](https://openhab.org/docs/configuration/transformations.html#script-transformation) for more general information on the usage of script transformations.

Reviewer 2: I created https://github.com/openhab/openhab-docs/pull/2488 so it will be handled in the documentation build. :slightly_smiling_face:

logger.info( historicItem.getState().toString() );
```

### Using scope

Simple usage of jsr223 scope objects

```python
from openhab import Registry

from scope import ON

Registry.getItem("Item1").sendCommand(ON)
```

### Logging

There are 3 ways of logging.

1. using normal print statements. In this case they are redirected to the default openhab logfile and marked with log level INFO or ERROR

```python
import sys

print("log message")

print("error message", file=sys.stderr)

```

2. using the logging module. Here you get a logging object, already initialized with the prefix "org.openhab.automation.pythonscripting"

```python
from openhab import logging

logging.info("info message")

logging.error("error message")
```

3. using the rule based logging module. Here you get a logging object, already initialized with the prefix "org.openhab.automation.pythonscripting.<RuleClassName>"

```python
from openhab import rule
from openhab.triggers import GenericCronTrigger

@rule( triggers = [ GenericCronTrigger("*/5 * * * * ?") ] )
class Test:
    def execute(self, module, input):
        self.logger.info("Rule was triggered")
```

## Decorators

### decorator @rule

> the decorator will register the decorated class as a rule. It will wrap and extend the class with the following functionalities

Author: is fixed

@when("Member of gTest_Switches received update")

@when("Item Test_Switch_1 received command")
@when("Item Test_Switch_2 received command OFF")

@when("Thing hue:device:default:lamp1 received update ONLINE")

@when("Thing hue:device:default:lamp1 changed from ONLINE to OFFLINE")

@when("Channel hue:device:default:lamp1:color triggered START")

@when("System started")
@when("System reached start level 50")

@when("Time cron 55 55 5 * * ?")
@when("Time is midnight")
@when("Time is noon")

@when("Time is 10:50")

@when("Datetime is Test_Datetime_1")
@when("Datetime is Test_Datetime_2 time only")

@when("Item added")
@when("Item removed")
@when("Item updated")

@when("Thing added")
@when("Thing removed")
@when("Thing updated")
```

### decorator @onlyif

```python
@onlyif("Item Test_Switch_2 equals ON")
@onlyif("Today is a holiday")
@onlyif("It's not a holiday")
@onlyif("Tomorrow is not a holiday")
@onlyif("Today plus 1 is weekend")
@onlyif("Today minus 1 is weekday")
@onlyif("Today plus 3 is a weekend")
@onlyif("Today offset -3 is a weekend")
@onylyf("Today minus 3 is not a holiday")
@onlyif("Yesterday was in dayset")
@onlyif("Time 9:00 to 14:00")
```

## Modules

### module scope

> The scope module encapsulates all [default jsr223 objects/presents](https://www.openhab.org/docs/configuration/jsr223.html#default-preset-importpreset-not-required) into a new object. You can use it like below

Author: same arguments as in my first comment related to {{base}} usage

```python
from scope import * # this makes all jsr223 objects available

print(ON)
```

```python
from scope import ON, OFF # this imports specific jsr223 objects

print(ON)
```

```python
import scope # this imports just the module

print(scope.ON)
```

> You can also import additional [jsr223 presents](https://www.openhab.org/docs/configuration/jsr223.html#rulesimple-preset) like

Author: same arguments as in my first comment related to {{base}} usage

```python
from scope import RuleSimple
from scope import RuleSupport
from scope import RuleFactories
from scope import ScriptAction
from scope import cache
from scope import osgi
```

> Additionally you can import all java classes from 'org.openhab' package like

Author: fixed

### class Timer 

| Function                 | Usage                                                                                 | Description                                                                                         |
| ------------------------ | ------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------- |
| createTimeout            | createTimeout(duration, callback, args=[], kwargs={}, old_timer = None, max_count = 0 ) | Create a timer that will run callback with arguments args and keyword arguments kwargs, after duration seconds have passed. If old_timer from e.g previous call is provided, it will be stopped if not already triggered. If max_count together with old_timer is provided, then 'max_count' times the old timer will be stopped and recreated, before the callback will be triggered immediately |

### class Set

This is a helper class which makes it possible to use a python 'set' as an argument for java class method calls

## Others

> ### python <=> java conversion

Author: fixed

> # Python 3 Scripting

Author: is fixed.

I was thinking first to permanently name it python 3, to be prepared for the future :-)

class Test:
    def execute(self, module, input):
        self.logger.info("Rule was triggered")
```

::: tip Note
By default, the scope, Registry and logger is automatically imported for UI based rules
:::
 

## `PY3` Transformation

openHAB provides several [data transformation services](https://www.openhab.org/addons/#transform) as well as the script transformations, that are available from the framework and need no additional installation.
It allows transforming values using any of the available scripting languages, which means Python Scripting is supported as well.
See the [transformation docs](https://openhab.org/docs/configuration/transformations.html#script-transformation) for more general information on the usage of script transformations.

Use Python Scripting as script transformation by:

1. Creating a script in the `$OPENHAB_CONF/transform` folder with the `.py` extension.
   The script should take one argument `input` and return a value that supports `toString()` or `null`:

   ```python
   def calc(input):
       # Do some data transformation here, e.g.
       return "String has" + data.length + "characters";
   calc(input)
   ```

> 2. Using `PY3(<scriptname>.py):%s` as Item state transformation.
> 3. Passing parameters is also possible by using a URL like syntax: `PY3(<scriptname>.py?arg=value)`.
>    Parameters are injected into the script and can be referenced like variables.
> 
> Simple transformations can aso be given as an inline script: `PY3(|...)`, e.g. `PY3(|"String has " + input.length + "characters")`.

Reviewer 2: ```suggestion
2. Using `PY3(<scriptname>.py):%s` as Item state transformation.
3. Passing parameters is also possible by using a URL like syntax: `PY3(<scriptname>.py?arg=value)`.
   Parameters are injected into the script and can be referenced like variables.

Simple transformations can also be given as an inline script: `PY3(|...)`, e.g. `PY3(|"String has " + input.length + "characters")`.
```

> Simple transformations can aso be given as an inline script: `PY3(|...)`, e.g. `PY3(|"String has " + input.length + "characters")`.

Author: So what is your recommendation here? I like to have PY for the new one and PY2 for the old one.

My recommendation is to keep it like it is for now.

And if this binding is merged. I will create a new pull request where I change the jython one to PY2 and PY for this addon.

> Simple transformations can aso be given as an inline script: `PY3(|...)`, e.g. `PY3(|"String has " + input.length + "characters")`.

Reviewer 2: Yes it's better to do this in a follow up PR so CI only builds this add-on when commits are added instead of all add-ons.

@rule(
    triggers = [
        ItemStateUpdateTrigger("Item1"),
        ItemCommandTrigger("Item1", scope.ON)
    ],
    conditions = [
        EphemerisCondition("notholiday")
    ]
)
class Test4:
    def execute(self, module, input):
        if Registry.getItem("Item2").postUpdateIfDifferent(scope.OFF):
            self.logger.info("Item2 was updated")
```
 
### Query thing status info

```python
from openhab import logger, Registry

info = Registry.getThing("zwave:serial_zstick:512").getStatusInfo()
logger.info(info.toString());
```

### Query historic item

```python
from openhab import logger, Registry
from datetime import datetime

historicItem = Registry.getItem("Item1").getPersistence().persistedState( datetime.now().astimezone() )
logger.info( historicItem.getState().toString() );

historicItem = Registry.getItem("Item2").getPersistence("jdbc").persistedState( datetime.now().astimezone() )
logger.info( historicItem.getState().toString() );
```

### Using scope

Simple usage of jsr223 scope objects

```python
from openhab import Registry

from scope import ON

Registry.getItem("Item1").sendCommand(ON)
```

### Logging

There are 3 ways of logging.

> 1. using normal print statements. In this case they are redirected to the default openhab logfile and marked with log level INFO or ERROR

Reviewer 2: ```suggestion
1. using normal print statements. In this case they are redirected to the default openHAB logfile and marked with log level INFO or ERROR
```

Author: is fixed

| Ping                     | see [openhab Ping api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/ping)        |                                                                          |
| ScriptExecution          | see [openhab ScriptExecution api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/scriptexecution) |                                                           |
| Semantic                 | see [openhab Semantic api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/semantic) |                                                                         |
| ThingAction              | see [openhab ThingAction api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/things) |                                                                   |
| Transformation           | see [openhab Transformation api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/transformation) |                                                             |

> | Voice                    | see [openhab Voice api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/voice)      |                                                                          |

Author: is fixed

| postUpdateIfDifferent    | postUpdateIfDifferent(state)                                                          |                                                                                                     |
| sendCommand              | sendCommand(command)                                                                  |                                                                                                     |
| sendCommandIfDifferent   | sendCommandIfDifferent(command)                                                       |                                                                                                     |
| getPersistence           | getPersistence(service_id = None)                                                     | [ItemPersistence](#class-itempersistence)                                         |
| getSemantic              | getSemantic()                                                                         | [ItemSemantic](#class-itemsemantic)                                               |

> | <...>                    | see [openHAB Item api](https://www.openhab.org/javadoc/latest/org/openhab/core/items/item) |                                                                                                |

Reviewer 2: ```suggestion
| <...>                    | see [openHAB Item API](https://www.openhab.org/javadoc/latest/org/openhab/core/items/item) |                                                                                                |
```

Author: is fixed

### class GroupItem 

GroupItem is an extended [Item](#class-item) which wraps results from getAllMembers & getMembers into [Items](#class-item)

### class ItemPersistence 

ItemPersistence is a wrapper around [openHAB PersistenceExtensions](https://www.openhab.org/javadoc/latest/org/openhab/core/persistence/extensions/persistenceextensions). The parameters 'item' and 'serviceId', as part of the Wrapped Java API, are not needed, because they are inserted automatically.

| Function                 | Usage                                                                                 | Description                                                                                         |
| ------------------------ | ------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------- |

> | getStableMinMaxState     | getStableMinMaxState(time_slot, end_time = None)                                      | Average calculation which takes into account the values ​​depending on their duration                 |
> | getStableState           | getStableState(time_slot, end_time = None)                                            | Average calculation which takes into account the values ​​depending on their duration                 |

Reviewer 2: Why the zero width spaces?

```suggestion
| getStableMinMaxState     | getStableMinMaxState(time_slot, end_time = None)                                      | Average calculation which takes into account the values depending on their duration                 |
| getStableState           | getStableState(time_slot, end_time = None)                                            | Average calculation which takes into account the values depending on their duration                 |
```

Author: ![grafik](https://github.com/user-attachments/assets/f09bb80b-55bf-47d2-924b-afb959f7f796)

it is just to align table column separator which makes it easier to read during edit.

| ------------------------ | ------------------------------------------------------------------------------------- |
| <...>                    | see [openHAB Semantics api](https://www.openhab.org/javadoc/latest/org/openhab/core/model/script/actions/semantics) |

### class Thing 

Thing is a wrapper around [openHAB Thing](https://www.openhab.org/javadoc/latest/org/openhab/core/thing/thing). 

| Function                 | Usage                                                                                 |
| ------------------------ | ------------------------------------------------------------------------------------- |
| <...>                    | see [openHAB Thing api](https://www.openhab.org/javadoc/latest/org/openhab/core/thing/thing) |

### class Channel 

Channel is a wrapper around [openHAB Channel](https://www.openhab.org/javadoc/latest/org/openhab/core/thing/type/channelgrouptype). 

| Function                 | Usage                                                                                 |
| ------------------------ | ------------------------------------------------------------------------------------- |

> | <...>                    | see [openHAB Channel api](https://www.openhab.org/javadoc/latest/org/openhab/core/thing/type/channelgrouptype) |

Author: is fixed

### class Timer 

| Function                 | Usage                                                                                 | Description                                                                                         |
| ------------------------ | ------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------- |
| createTimeout            | createTimeout(duration, callback, args=[], kwargs={}, old_timer = None, max_count = 0 ) | Create a timer that will run callback with arguments args and keyword arguments kwargs, after duration seconds have passed. If old_timer from e.g previous call is provided, it will be stopped if not already triggered. If max_count together with old_timer is provided, then 'max_count' times the old timer will be stopped and recreated, before the callback will be triggered immediately |

### class Set

> This is a helper class which makes it possible to use a python 'set' as an argument for java class method calls

Author: is fixed

## Others

### Threading

Thread or timer objects which was started by itself should be registered in the lifecycleTracker to be cleaned during script unload.

```python
import scope
import threading

class Timer(theading.Timer):
    def __init__(self, duration, callback):
        super().__init__(duration, callback)

    def shutdown(self):
        if not self.is_alive():
            return
        self.cancel()
        self.join()

def test():
    print("timer triggered")

job = Timer(60, test)
job.start()

scope.lifecycleTracker.addDisposeHook(job.shutdown)
```

Timer objects created via `openhab.Timer.createTimeout`, however, automatically register in the disposeHook and are cleaned on script unload.

```python
from openhab import Timer

def test():
    print("timer triggered")

Timer.createTimeout(60, test)
```

> Below is a complex example of 2 sensor values ​​that are expected to be transmitted in a certain time window (e.g. one after the other).

Reviewer 2: More zero width spaces?

```suggestion
Below is a complex example of 2 sensor values that are expected to be transmitted in a certain time window (e.g. one after the other).
```

@rule(
    triggers = [
        ItemStateChangeTrigger("Room_Temperature_Value"),
        ItemStateChangeTrigger("Room_Humidity_Value")
    ]
)
class UpdateInfo:
    def __init__(self):
        self.update_timer = None
    
    def updateInfoMessage(self):
        msg = "{}{} °C, {} %".format(Registry.getItemState("Room_Temperature_Value").format("%.1f"), Registry.getItemState("Room_Temperature_Value").format("%.0f"))
        Registry.getItem("Room_Info").postUpdate(msg)
        self.update_timer = None

    def execute(self, module, input):
        self.update_timer = Timer.createTimeout(5, self.updateInfoMessage, old_timer = self.update_timer, max_count=2 )

```

### Python <=> Java conversion

Conversion occurs in both directions

| Python class              | Java class    |
| ------------------------- | ------------- |
| datetime with timezone    | ZonedDateTime |
| datetime without timezone | Instant       |
| timedelta                 | Duration      |
| list                      | Collection    |
| Set(set)                  | Set           |
| Item                      | Item          |

### typical log errors

#### Exception during helper lib initialisation

> There were problems during the deployment of the helper libs. A typical error is an insufficient permission. The folder "conf/automation/python/" must be writeable by openhab.

Author: is fixed

#### Failed to inject import wrapper

The reading the python source file "conf/automation/python/lib/openhab/__wrapper__.py" failed.

This could either a permission/owner problem or a problem during deployment of the helper libs. You should check that this file exists and it is readable by openhab. You should also check your logs for a message related to the helper lib deployment by just grep for "helper lib"

### limitations

> - graalby can't handle arguments in constructors of java objects. Means you can't instantiate a javaobject in python with a parameter. https://github.com/oracle/graalpython/issues/367
> - graalpy does not really support python 'set' types as arguments of function calls to java objects https://github.com/oracle/graalpython/issues/260

Reviewer 2: ```suggestion
- GraalPy can't handle arguments in constructors of java objects. Means you can't instantiate a Java object in Python with a parameter. https://github.com/oracle/graalpython/issues/367
- GraalPy does not really support python 'set' types as arguments of function calls to Java objects https://github.com/oracle/graalpython/issues/260
```

Author: is fixed
